x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlAnomaliesIndexUpdate.java

```diff
@@ -0,0 +1,235 @@
+import org.elasticsearch.xpack.core.ml.utils.MlIndexAndAlias;
+import org.elasticsearch.xpack.core.ml.utils.MlStrings;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+/**
+ * Rollover the various .ml-anomalies result indices
+ * updating the read and write aliases
+ */
+public class MlAnomaliesIndexUpdate implements MlAutoUpdateService.UpdateAction {
+
+    private static final Logger logger = LogManager.getLogger(MlAnomaliesIndexUpdate.class);
+
+    private final IndexNameExpressionResolver expressionResolver;
+    private final OriginSettingClient client;
+
+    public MlAnomaliesIndexUpdate(IndexNameExpressionResolver expressionResolver, Client client) {
+        this.expressionResolver = expressionResolver;
+        this.client = new OriginSettingClient(client, ML_ORIGIN);
+    }
+
+    @Override
+    public boolean isMinTransportVersionSupported(TransportVersion minTransportVersion) {
+        // Automatic rollover does not require any new features
+        // but wait for all nodes to be upgraded anyway
+        return minTransportVersion.onOrAfter(TransportVersions.ML_ROLLOVER_LEGACY_INDICES);
+    }
+
+    @Override
+    public boolean isAbleToRun(ClusterState latestState) {
+        // Find the .ml-anomalies-shared and all custom results indices
+        String[] indices = expressionResolver.concreteIndexNames(
+            latestState,
+            IndicesOptions.lenientExpandOpenHidden(),
+            AnomalyDetectorsIndex.jobResultsIndexPattern()
+        );
+
+        for (String index : indices) {
+            IndexRoutingTable routingTable = latestState.getRoutingTable().index(index);
+            if (routingTable == null || routingTable.allPrimaryShardsActive() == false) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public String getName() {
+        return "ml_anomalies_index_update";
+    }
+
+    @Override
+    public void runUpdate(ClusterState latestState) {
+        List<Exception> failures = new ArrayList<>();
+
+        // list all indices starting .ml-anomalies-
+        // this includes the shared index and all custom results indices
+        String[] indices = expressionResolver.concreteIndexNames(
+            latestState,
+            IndicesOptions.lenientExpandOpenHidden(),
+            AnomalyDetectorsIndex.jobResultsIndexPattern()
+        );
+
+        for (String index : indices) {
+            boolean isCompatibleIndexVersion = MlIndexAndAlias.indexIsReadWriteCompatibleInV9(
+                latestState.metadata().index(index).getCreationVersion()
+            );
+
+            if (isCompatibleIndexVersion) {
+                continue;
+            }
+
+            PlainActionFuture<Boolean> updated = new PlainActionFuture<>();
+            rollAndUpdateAliases(latestState, index, updated);
+            try {
+                updated.actionGet();
```

Comment:
Should we add a timeout here? Otherwise this could block forever right?

I feel like I remember reading something about `PlainActionFuture` being unsafe? Maybe I'm making that up. I know we use it in tests all the time but wasn't sure if it was common to use it outside of that.

Reply:
Patrick made the same point on this PR https://github.com/elastic/elasticsearch/pull/120405

I'm following the pattern established in https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedConfigAutoUpdater.java#L112  but that is not necessarily a good thing especially given I've had 2 sets of feedback on this.  I will try to rewrite it non-blocking